pilot/platform/kube/deregister.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"strings"
+
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/istio/pkg/log"
+)
+
+// VerifyEndpoint verifies if the provided IP to deregister has already
+// been registered.
+func VerifyEndpoint(eps *v1.Endpoints, ip string) bool {
+	var match bool
+	subsetLen := len(eps.Subsets)
+	/*
+		1. Check if this is the only entry; if this is the only entry
+		   then delete the entry and should return an empty subset.
+		2. Check if service endpoint is part of only one subset. If so
+		   wipe off the empty subset and return non-empty subsets.
+		3. Check if service endpoint is part of multiple subset. If so
+		   wipe off the empty subset and coalease non-empty subsets.
+	*/
+	for i, ss := range eps.Subsets {
+		for j, ipaddr := range ss.Addresses {
+			if strings.Compare(ipaddr.IP, ip) == 0 {
+				/*
+					If the IP address match then remove that
+					specific IP address.
+				*/
+				match = true
+				if j < len(ss.Addresses) {
+					ss.Addresses = append(ss.Addresses[:j], ss.Addresses[j+1:]...)
+				} else {
+					ss.Addresses = ss.Addresses[:j]
+				}
+			}
+		}
+		if len(ss.Addresses) == 0 {
+			if len(eps.Subsets) < subsetLen {
+				eps.Subsets = append(eps.Subsets[:len(eps.Subsets)], eps.Subsets[len(eps.Subsets)+1:]...)
+			} else {
+				eps.Subsets = append(eps.Subsets[:i], eps.Subsets[i+1:]...)
+			}
+		} else {
+			if len(eps.Subsets) < subsetLen {
+				eps.Subsets[len(eps.Subsets)-1].Addresses = ss.Addresses
+			} else {
+				eps.Subsets[i].Addresses = ss.Addresses
```

Comment:
only this line should be needed for update

Reply:
Lets consider a case where we want to deregister ip "10.128.0.18" which is present in 1st and 2nd entry whereas it is not present in the 3rd entry. This use case handles the coaleasing case we discussed during the design phase. 
 
Subsets: []v1.EndpointSubset{
                                        {
                                                []v1.EndpointAddress{{"10.128.0.17", "", nil, nil}, {"10.128.0.18", "", nil, nil}, {"10.128.0.19", "", nil, nil}},
                                                []v1.EndpointAddress{{"", "", nil, nil}},
                                                []v1.EndpointPort{{"3309", 3309, "TCP"}},
                                        },
                                        {
                                                []v1.EndpointAddress{{"10.128.0.18", "", nil, nil}},
                                                []v1.EndpointAddress{{"", "", nil, nil}},
                                                []v1.EndpointPort{{"3307", 3307, "TCP"}},
                                        },
                                        {
                                                []v1.EndpointAddress{{"10.128.0.20", "", nil, nil}, {"10.128.0.21", "", nil, nil}, {"10.128.0.22", "", nil, nil}},
                                                []v1.EndpointAddress{{"", "", nil, nil}},
                                                []v1.EndpointPort{{"3308", 3308, "TCP"}},
                                        },
                                },

The way the algorithm behaves is when its sees a match for "10.128.0.18" in 1st entry it will change []v1.EndpointAddress{{"10.128.0.17", "", nil, nil}, {"10.128.0.18", "", nil, nil}, {"10.128.0.19", "", nil, nil}}, => 
[]v1.EndpointAddress{{"10.128.0.17", "", nil, nil}, {"10.128.0.19", "", nil, nil}}, 

For 2nd entry it will remove the subset as there is only one entry

 []v1.EndpointAddress{{"10.128.0.18", "", nil, nil}},
 []v1.EndpointAddress{{}},

so we now have only 1st entry and 3rd entry so subset length becomes 2 not 3 what it was initially. If this check is not added then we will have errors with referencing out of bound array. 

Please let me know if you still think we don't need it.

Reply:
that's a classic modifying the list you iterate upon, the solution is to use the iterator to mutate the entry (not the index) ?